internal/provider/providers/vultr/provider.go

```diff
@@ -0,0 +1,195 @@
+	}
+	return nil
+}
+
+func (p *Provider) String() string {
+	return utils.ToString(p.domain, p.owner, constants.Vultr, p.ipVersion)
+}
+
+func (p *Provider) Domain() string {
+	return p.domain
+}
+
+func (p *Provider) Owner() string {
+	return p.owner
+}
+
+func (p *Provider) IPVersion() ipversion.IPVersion {
+	return p.ipVersion
+}
+
+func (p *Provider) IPv6Suffix() netip.Prefix {
+	return p.ipv6Suffix
+}
+
+func (p *Provider) Proxied() bool {
+	return false
+}
+
+func (p *Provider) BuildDomainName() string {
+	return utils.BuildDomainName(p.owner, p.domain)
+}
+
+func (p *Provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.BuildDomainName(), p.BuildDomainName()),
+		Owner:     p.Owner(),
+		Provider:  fmt.Sprintf("<a href=\"https://my.vultr.com/dns/edit/?domain=%s#dns-records\">Vultr</a>", p.domain),
+		IPVersion: p.ipVersion.String(),
+	}
+}
+
+func (p *Provider) setHeaders(request *http.Request) {
+	headers.SetUserAgent(request)
+	headers.SetContentType(request, "application/json")
+	headers.SetAccept(request, "application/json")
+	headers.SetAuthBearer(request, p.token)
+}
+
+func (p *Provider) Update(ctx context.Context, client *http.Client, ip netip.Addr) (newIP netip.Addr, err error) {
+	recordType := constants.A
+	if ip.Is6() {
+		recordType = constants.AAAA
+	}
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   "api.vultr.com",
+		Path:   fmt.Sprintf("/v2/domains/%s/records", p.domain),
+	}
+
+	requestData := struct {
+		Type string `json:"type"`
+		IP   string `json:"data"`
+		Name string `json:"name"`
+		TTL  uint32 `json:"ttl"`
+	}{
+		Type: recordType,
+		IP:   ip.String(),
+		Name: p.owner,
+		TTL:  p.ttl,
+	}
+
+	buffer := bytes.NewBuffer(nil)
+	encoder := json.NewEncoder(buffer)
+	err = encoder.Encode(requestData)
+	if err != nil {
+		return netip.Addr{}, fmt.Errorf("json encoding request data: %w", err)
+	}
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), buffer)
+	if err != nil {
+		return netip.Addr{}, fmt.Errorf("creating http request: %w", err)
+	}
+	p.setHeaders(request)
+
+	response, err := client.Do(request)
+	if err != nil {
+		return netip.Addr{}, err
+	}
+	defer response.Body.Close()
+
+	decoder := json.NewDecoder(response.Body)
+	var parsedJSON struct {
+		Error string
+		Status uint32
+		Record struct {
+			ID   string `json:"id"`
+			Type string `json:"type"`
+			IP   string `json:"data"`
+			Name string `json:"name"`
+			TTL  uint32 `json:"ttl"`
+		}
+	}
+	err = decoder.Decode(&parsedJSON)
+	if err != nil {
+		return netip.Addr{}, fmt.Errorf("json decoding response body: %w", err)
+	}
+
+	if parsedJSON.Error != "" {
+		return netip.Addr{}, fmt.Errorf("API Error: %s", parsedJSON.Error)
+	}
+
+	if response.StatusCode != http.StatusCreated {
+		return netip.Addr{}, fmt.Errorf("%w: %d: %s",
+			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.BodyToSingleLine(response.Body))
+	}
```

Comment:
I feel like this code would never trigger because we check the parsed error before it. Maybe have something like this instead:

```go
	if response.StatusCode != http.StatusCreated {
		return netip.Addr{}, fmt.Errorf("%w: %d: %s",
			errors.ErrHTTPStatusNotValid, response.StatusCode, parsedJSON.Error)
	}

	if parsedJSON.Error != "" {
		return netip.Addr{}, fmt.Errorf("%w: %s", errors.ErrUnsuccessful, parsedJSON.Error)
	}
```

Also for your own knowledge, note the previously used `utils.BodyToSingleLine(response.Body)` would not work because it has already been read by the JSON decoder. One way to work around this is to io.ReadAll(response.Body) to have it in a byte slice, then use `json.Unmarshal(byteSlice, &parsedJSON)` and you can then re-use `byteSlice` after for example with `utils.ToSingleLine(string(byteSlice))`. Not something to do here really I think though 😄

Reply:
I can't flip it as that would return, their api returns an error string if it's not 201, so that checks for the error message and displays it for the user. If the object doesn't contain an error though, we can rely on the status code. 

Thanks for your tip

Reply:
There are 3 cases
1. 201 created
2. HTTP error (400, 401, 403, 404) with error message
3. No documentation about 5xx where probably there won't be an error message

So I first check the error message if it exists, and if not, I check the status code. Does that make sense?

Reply:
Ah interesting, so the code needs more change then.

I suggest you have a little function:

```go
func parseJSONError(body io.Reader) (message string) {
	var parsedJSON struct {
		Error string `json:"error"`
	}
	decoder := json.NewDecoder(body)
	err := decoder.Decode(&parsedJSON)
	if err != nil {
		return "<json decoding error: " + err.Error() + ">"
	}
	return parsedJSON.Error
}
```

and then have a switch on the status code:


```go
	defer response.Body.Close()
	
	switch response.StatusCode {
	case http.StatusCreated:
	case http.StatusBadRequest:
		return fmt.Errorf("%w: %s", errors.ErrBadRequest, parseJSONError(response.Body))
	case http.StatusUnauthorized, http.StatusForbidden:
		return fmt.Errorf("%w: %s", errors.ErrAuth, parseJSONError(response.Body))
	case http.StatusNotFound:
		return fmt.Errorf("%w: %s", errors.ErrDomainNotFound, parseJSONError(response.Body))
	default:
		return netip.Addr{}, fmt.Errorf("%w: %s: %s",	errors.ErrHTTPStatusNotValid,
			response.Status, utils.BodyToSingleLine(response.Body))
	}

	decoder := json.NewDecoder(response.Body)
	var parsedJSON struct {
		Error string
		Status uint32
		Record struct {
			ID   string `json:"id"`
			Type string `json:"type"`
			IP   string `json:"data"`
			Name string `json:"name"`
			TTL  uint32 `json:"ttl"`
		}
	}
	err = decoder.Decode(&parsedJSON)
	if err != nil {
		return netip.Addr{}, fmt.Errorf("json decoding response body: %w", err)
	}
```